include/s3select_functions.h

```diff
@@ -151,18 +170,27 @@ class __function : public base_statement
   }
 
 
-  std::vector<base_statement*> get_arguments()
+  bs_stmt_vec_t& get_arguments()
   {
     return arguments;
   }
 
-  virtual ~__function()
-  {
-    arguments.clear();
-  }
+  virtual ~__function() {}
 };
 
 
+  void s3select_functions::clean()
+  {
+    for(auto d : __all_query_functions)
+    {
+      if (d->is_function())
+      {
+        dynamic_cast<__function*>(d)->impl()->__call_destructor();
+      }
+      d->__call_destructor();
```

Comment:
while I'm looking at this code section, I see a bunch of member variables whose names begin with double-underscore;  I don't think that is helpful.  Just give member variables ordinary names.